learn/quickstart/derived_types.md

```diff
@@ -0,0 +1,299 @@
+
+    !5th case: coarray
+    real, allocatable, codimension[:] :: z(:) ! a 1d_dynamic array shared in all threads. Coarrays will be discussed in the Advanced programming mini-book.
+    !or 
+    real, allocatable :: z(:)[:] ! here the [:] is the [attr-dependent-spec] and implies the codimension[:].
+
+    !6th case: contiguous
+    real, contiguous, pointer :: x(:)
+end type
+```
+
+{% include note.html content="In the above example the cases 4, 5 and 6 make use of Fortran `pointer` and `coarray` features that have not been addressed in this quickstart tutorial. However, they are presented here, in order for the readers to know that these features do exist and be able to recognize them. These features will be covered in detail in the upcoming `Advanced programing` mini-book." %}
+
+## Type-bound procedures
+
+A derived type is possible to contain procedures either `functions` or `subroutines` that are **bound** to this derived type. Type procedures must follow the `contains` statement that, in return, must be used within the derived type and after all [member-variables] have been declared. 
+
+{% include note.html content="It is impossible to describe type-bound procedures in their full syntax without delving into OOP features of modern Fortran. For that reason only a simple example is provided in this final part, to demonstrate a very basic use." %}
+
+Example of a derived type with basic bound-procedure:
+
+```fortran
+module mymod
+implicit none
+private
+public t_square
+
+type :: t_square
+    real :: side
+    contains
+        procedure :: area !procedure declaration
+end type
+
+contains
+    ! procedure definition
+    real function area(self) result(res)
+        class(t_square), intent(in) :: self 
+        res = self%side * self%side
+    end function
+end module
+
+program main
+use mymod
+implicit none
+! variables declaration
+type(t_square) :: sq
+real :: x, side
+
+! variables initialization
+side    = 0.5
+sq%side = side
+
+x       = sq%area() ! self does not appear here, it has been passed implicitly
+! do stuff with x...
+end program
+```
+What is new:
+
+ - **self** is a random name that was chosen to represent the derived type t_square that is passed as an argument to the bound-function in order to have access to its data-members. By passing it like that it is ensured that later during its use the t_square will be passed automatically and not by the client.
+ - in order to have the above functionality the new keyword `class` replaced the `type` one. With `class` the OOP feature *polymorphism* is introduced. 
+ - since the bound-procedure **area** was defined as a function it cannot be called by itself, it can only appear as **rhs** object, that is why it is used like `x = sq%area()`. The 'stand-alone' functionality is covered by a subroutine, and the above example should be modified like:
```

Comment:
The 3rd bullet actually didn't belong in the "what's new" list, but is its own separate lesson. So I put it in a new paragraph.

```suggestion

In the above example, the type-bound procedure **area** is defined as a function and can be invoked only in an expression, for example `x = sq%area()` or `print *, sq%area()`. If you define it instead as a subroutine, you can invoke it from its own `call` statement:
```